AD_query_tool.py

```diff
@@ -78,22 +78,61 @@ def enabled_inactive_users():
             last_logon_datetime = to_date_time(timestamp) 
             if last_logon_datetime <= last_month_datetime:
                 enabled_inactive_users_count += 1
-        elif timestamp == 0:
+        else:
             enabled_inactive_users_count += 1  
-    return(enabled_inactive_users_count)        
+    return(enabled_inactive_users_count)   
+
+# When PASSWD_NOTREQD flag is set, no password is required
+def enabled_accounts_without_password():   
+    df = pd.DataFrame(df_including_password_age())    
+    df_enabled_accounts_without_password = df.loc[df['UserAccountControl'] & 0x0020 == 1] 
+    df_enabled_accounts_without_password_amount = df_enabled_accounts_without_password['CN'].count()
+    return(df_enabled_accounts_without_password_amount)
+
+def enabled_domain_admins():   
+    df = pd.DataFrame(df_including_password_age())
+    df_enabled_users = enabled_users(df)
+    search_list = ['Domain Admins', 'Administrator', 'Enterprise Admins']
+    df_domain_admins = df_enabled_users.loc[df_enabled_users['GroupsMembership'].str.contains('|'.join(search_list))]
+    df_domain_admins_amount = df_domain_admins['CN'].count()
+    return(df_domain_admins)
+
+def enabled_domain_admins_avg_password_age():
+    df_enabled_domain_admins = enabled_domain_admins()
+    avg_password_age = df_enabled_domain_admins['PasswordAge'].mean()
+    return(int(avg_password_age))
+
+def enabled_domain_admins_amount():   
+    df_domain_admins = enabled_domain_admins()
+    enabled_domain_admins_amount = df_domain_admins['CN'].count()
+    return(enabled_domain_admins_amount)
+
+# When SMARTCARD_REQUIRED flag is set, it forces the user to log on by using a smart card
+def enabled_admins_card_req():   
+    df_domain_admins = enabled_domain_admins()
+    admins_card_req = df_domain_admins.loc[df_domain_admins['UserAccountControl'] & 0x40000 == 1]
+    admins_card_req_amount = admins_card_req['CN'].count()
+    return(admins_card_req_amount)
+
+# When NOT_DELEGATED flag is set, the security context of the user isn't delegated to a service even if the service account is set as trusted for Kerberos delegation
+def enabled_admins_account_not_delegated():   
+    df_domain_admins = enabled_domain_admins()
+    admins_non_delegated = df_domain_admins.loc[df_domain_admins['UserAccountControl'] & 0x100000 == 1]
+    admins_non_delegated_amount = admins_non_delegated['CN'].count()
+    return(admins_non_delegated_amount)
 
 def main():
```

Comment:
your main function looks awesome now, clean and essential :)

Reply:
:)